Donation-platform/src/main/java/com/google/DbCalls.java

```diff
@@ -110,17 +97,144 @@ public void addUser(String userName, String email,
+        .setFilter(new FilterPredicate("email", FilterOperator.EQUAL, email));
+    PreparedQuery pq = datastore.prepare(query);
+    Entity result = pq.asSingleEntity();
+    Users userResult = setUsersClass(result);
+    return userResult;
+  }
+
+  // Function turns entity into class.
+  public Charity setCharityClass(Entity entity){
+    Key id = (Key) entity.getKey();
+    String name = (String) entity.getProperty("name");
+    String link = (String) entity.getProperty("link");
+    List<Key> categories = (List<Key>) entity.getProperty("categories");
+    String description = (String) entity.getProperty("description");
+    Double trendingScore = (Double) entity.getProperty("trendingScore");
+    // Converts data to charity object.
+    Charity charityData = new Charity(id, name, link, categories, description, trendingScore);
+    return charityData;
+  }
+  // Function turns entity into class.
+  public Tag setTagClass(Entity entity){
+    Key id = (Key) entity.getKey();
+    String name = (String) entity.getProperty("name");
+    Double trendingScore = (Double) entity.getProperty("trendingScore");
+    // Converts data to tag object.
+    Tag tagData = new Tag(id, name, trendingScore);
+    return tagData;
+  }
+  // Function turns entity into class.
+  public Users setUsersClass(Entity entity){
+    Key id = (Key) entity.getKey();
+    String userName = (String) entity.getProperty("userName");
+    String email = (String) entity.getProperty("email");
+    List<Key>  userInterests = (List<Key> ) entity.getProperty("userInterests");
+    List<Key>  charitiesDonatedTo = (List<Key> ) entity.getProperty("charitiesDonatedTo");
+    // Converts data to users object.
+    Users userData = new Users(id, userName, email, userInterests, charitiesDonatedTo);
+    return userData;
+  }
+  // Function takes modified class and updates database.
+  public void updateCharity(Charity charity) throws Exception{
+    Key id = charity.getId();
     Entity charityEntity = datastore.get(id);
-    charityEntity.setProperty("trendingScore", trendingScore);
+    charityEntity.setProperty("name", charity.getName());
+    charityEntity.setProperty("link", charity.getLink());
+    charityEntity.setProperty("categories", charity.getCategories());
+    charityEntity.setProperty("description", charity.getDescription());
+    charityEntity.setProperty("trendingScore", charity.getTrendingScoreCharity());
     datastore.put(charityEntity);
   }
-
+  // Function takes modified class and updates database.
+  public void updateTag(Tag tag) throws Exception{
+    Key id = tag.getId();
+    Entity tagEntity = datastore.get(id);
+    tagEntity.setProperty("name", tag.getName());
+    tagEntity.setProperty("trendingScore", tag.getTrendingScoreTag());
+    datastore.put(tagEntity);
+  }
+  // Function takes modified class and updates database.
+  public void updateUser(Users user) throws Exception{
+    Key id = user.getId();
+    Entity userEntity = datastore.get(id);
```

Comment:
since this is the only place we are using it, can we use it directly? instead of declaring it

Reply:
Please take care of this throughout the file

Reply:
sorry charu, i don't immediately understand the directive in this comment.
are you saying i shouldn't declare the entities?

Reply:
oh i read the article down below and it's clear now

Reply:
I can change it but if I call it each time I make a property change then would that not mean I am invoking a database call (.get()) each time. Is this good practice?

Reply:
Furthermore, on that note, I believe the use of the entity variable is justified in this case as we need the object to alter before returning it to the database. Calling it directly leads to a nullpointerexception and detracts from the readability. I believe per the datastore api docs, this was how it was modified usually.

Reply:
Sorry i didnt get that. So "Entity userEntity = datastore.get(id)" is being used only once when you do "datastore.put()", why is it everytime then? 

Regarding null pointer, even when you are declaring it as an object, I dont see any handling for Null Pointer. Irrespective of declaring it as a variable or directly using it we should handle null pointer.